src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Debugging/RazorBreakpointSpanEndpoint.cs

```diff
@@ -0,0 +1,202 @@
+            ILoggerFactory loggerFactory)
+        {
+            if (projectSnapshotManagerDispatcher is null)
+            {
+                throw new ArgumentNullException(nameof(projectSnapshotManagerDispatcher));
+            }
+
+            if (documentResolver is null)
+            {
+                throw new ArgumentNullException(nameof(documentResolver));
+            }
+
+            if (documentVersionCache is null)
+            {
+                throw new ArgumentNullException(nameof(documentVersionCache));
+            }
+
+            if (documentMappingService is null)
+            {
+                throw new ArgumentNullException(nameof(documentMappingService));
+            }
+
+            if (loggerFactory is null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            _projectSnapshotManagerDispatcher = projectSnapshotManagerDispatcher;
+            _documentResolver = documentResolver;
+            _documentVersionCache = documentVersionCache;
+            _documentMappingService = documentMappingService;
+            _logger = loggerFactory.CreateLogger<RazorBreakpointSpanEndpoint>();
+        }
+
+        public async Task<RazorBreakpointSpanResponse?> Handle(RazorBreakpointSpanParams request, CancellationToken cancellationToken)
+        {
+            var info = await TryGetDocumentSnapshotAndVersionAsync(request.Uri.GetAbsoluteOrUNCPath(), cancellationToken).ConfigureAwait(false);
+            if (info is null)
+            {
+                return null;
+            }
+
+            var (documentSnapshot, documentVersion) = info;
+
+            var codeDocument = await documentSnapshot.GetGeneratedOutputAsync();
+            var sourceText = await documentSnapshot.GetTextAsync();
+            var linePosition = new LinePosition(request.Position.Line, request.Position.Character);
+            var hostDocumentIndex = sourceText.Lines.GetPosition(linePosition);
+
+            if (codeDocument.IsUnsupported())
+            {
+                return null;
+            }
+
+            var projectedIndex = hostDocumentIndex;
+            var languageKind = _documentMappingService.GetLanguageKind(codeDocument, hostDocumentIndex);
+            // If we're in C#, then map to the right position in the generated document
+            if (languageKind == RazorLanguageKind.CSharp &&
+                !_documentMappingService.TryMapToProjectedDocumentPosition(codeDocument, hostDocumentIndex, out _, out projectedIndex))
+            {
+                return null;
+            }
+            // Otherwise see if there is more C# on the line to map to
+            else if (languageKind == RazorLanguageKind.Html &&
+                !_documentMappingService.TryMapToProjectedDocumentOrNextCSharpPosition(codeDocument, hostDocumentIndex, out _, out projectedIndex))
+            {
+                return null;
+            }
+            else
```

Comment:
This intended?

Reply:
Yeah, just in case the language kind comes back as `Razor`

Reply:
Wait but doesn't this section always return `null` then making the code on lines 110+ unreachable?

Reply:
oh, i see.. oops.. thats what I get for making a last minute change in the web UI :)